types/node/vm.d.ts

```diff
@@ -56,6 +56,29 @@ declare module "vm" {
          */
         columnOffset?: number | undefined;
     }
+    type ImportModuleDynamically<
+        Mode extends "Script" | "Function" | "SourceTextModule" | "Context" = never,
+    > = Mode extends never ? unknown
+        :
+            | (
+                // When importModuleDynamically is `vm.constants.USE_MAIN_CONTEXT_DEFAULT_LOADER`
+                Mode extends "Script" | "Function" | "Context" ? typeof constants.USE_MAIN_CONTEXT_DEFAULT_LOADER
+                    : never
+            )
+            | (
+                // When importModuleDynamically is a function
+                (
+                    specifier: string,
+                    script: Mode extends "Script" ? Script
+                        : Mode extends "Function" ? ReturnType<typeof compileFunction>
+                        : Mode extends "SourceTextModule" ? SourceTextModule
+                        : Mode extends "Context" ? Context
+                        : never,
+                    importAttributes: ImportAttributes,
+                ) =>
+                    | Module
+                    | Promise<Module>
+            );
```

Comment:
This really doesn't seem to me to benefit from this level of conditional computation – this just adds unnecessary instantiation cost.

What would be better would be a simple alias for the loader, generic over the `referrer` argument, like so:
```ts
type DynamicModuleLoader<T> = (specifier: string, referrer: T, importAttributes: ImportAttributes) => Module | Promise<Module>;
```

which can then be used inline:

```ts
interface ScriptOptions extends BaseOptions {
    ...
    importModuleDynamically?: DynamicModuleLoader<Script> | typeof typeof constants.USE_MAIN_CONTEXT_DEFAULT_LOADER | undefined;
}
```
etc.